source/engines/rollEngine.js

```diff
@@ -1,548 +1,771 @@
-			return `${this.#resultset.toString(frac)} = ${this.#resultset.getResults()}/${this.#resultset.getMaxResults()}${this.#extraText ? ` ${this.#extraText}` : ''}`;
-		} else {
-			return `${this.#resultset.toString(frac)} = ${this.#resultset.getResults()}${this.#extraText ? ` ${this.#extraText}` : ''}`;
-		}
-	}
-}
-
-function parseRoll(input) { //This needs some wild revision before it'll work
-	input = input.replace(/\(/g,' ( ').replace(/\)/g,' ) ')
-	var splitInput = input.split(' ');
-	var rollArray = splitInput.reduce((acc, cur) => {
-		if (acc[acc.length-1] == '║' || cur == '') {
-			return acc;
-		} else if (/[\+\-\*]/.test(cur) || /([0-9]+|[0-9]+d[0-9]+([dk][lh]?[0-9]+)?)([\+\-\*]([0-9]+|[0-9]+d[0-9]+([dk][lh]?[0-9]+)?))*$/.test(cur) || /\(/.test(cur) || /\)/.test(cur)) {
-			acc.push(cur);
-			return acc;
-		} else { //stop on the first non-roll
-			acc.push('║');
-			return acc;
-		}
-	}, []);
-	rollArray[rollArray.length-1] == '║' ? rollArray.pop() : ()=>{};
-	var textArray = splitInput.reduceRight((acc, cur) => {
-		if (acc[0] == '║' || cur == '') {
-			return acc;
-		} else if (/([0-9]+|[0-9]+d[0-9]+([dk][lh]?[0-9]+)?)([\+\-\*]([0-9]+|[0-9]+d[0-9]+([dk][lh]?[0-9]+)?))*$/.test(cur) || /[0-9]+$/.test(cur) || /\)/.test(cur)) {
-			acc.unshift('║');
-			return acc;
-		} else {
-			acc.unshift(cur);
-			return acc;
-		}
-	}, []);
-	textArray[0] == '║' ? textArray.shift() : ()=>{};
-	return new ResultBundle(rollArray.join(''),textArray.join(' '));
-}
-
-function getRollString(input, frac = false, simple = false) {
-	return parseRoll(input).toString(frac, simple);
-}
-
-module.exports = {parseRoll, getRollString}
+class Result {
+	#result;
+
+	constructor(result) {
+		this.#result = result;
+	}
+
+	getResult() {
+		return this.#result;
+	}
+
+	toString() {
+		return `${this.#result}`;
+	}
+}
+
+class RollResult extends Result {
+	#dieSides;
+
+	constructor(dieSides) {
+		super(Math.floor(Math.random() * dieSides) + 1);
+		this.#dieSides = dieSides;
+	}
+
+	getMaxValue() {
+		return this.#dieSides;
+	}
+
+	toString(frac = false) {
+		return frac ? `${this.getResult()}/${this.getMaxValue()}` : `${this.getResult()}`;
+	}
+}
+
+class StringResult extends Result {
+	#optList;
+
+	constructor(result, optList) {
+		super(result);
+		this.#optList = optList;
+	}
+
+	getMaxValue() {
+		return this.getResult();
+	}
+
+	toString(frac = false) {
+		return frac ? `${this.getResult()} from the list ${this.#optList}` : this.getResult();
+	}
+}
+
+const two_ops = { // Mathematical operations we currently support for rolls that use two arguments
+	ADD: '+',
+	SUBTRACT: '-',
+	MULTIPLY: '*'
+}
+
+const keep_ops = { // Operations for keeping or dropping rolls of a result set
+	KEEP: 'k',
+	KEEP_HIGHEST: 'kh',
+	DROP_LOWEST: 'dl',
+	DROP: 'd',
+	DROP_HIGHEST: 'dh',
+	KEEP_LOWEST: 'kl'
+}
+
+class ResultSet {
+
+	constructor() {}
+
+	static #surroundedByParens(str, symbolPos) {
+		var rightParensUnenclosed = 0,
+			leftParensUnenclosed = 0;

```

Comment:
```suggestion
		let rightParensUnenclosed = 0,
			leftParensUnenclosed = 0;
```
These don't need to ignore block-scope.

Reply:
If I remember correctly, this is from old code where my beef with `let` was stronger. (It is still strong; it shouldn't be named `let`.) If anything, there should be a whole pass to switch `var` to `let` in this file as a separate task for consistency at minimum.